src/components/BulkEditModal/BulkEditModal.tsx

```diff
@@ -0,0 +1,222 @@
+    margin-bottom: ${theme.spacing.md};
+  `,
+  probesWrapper: css`
+    display: flex;
+    flex-direction: row;
+    flex-wrap: wrap;
+  `,
+  halfWidth: css`
+    width: 50%;
+  `,
+});
+
+const BulkEditModal = ({ onDismiss, onSuccess, onError, isOpen, selectedChecks, action, instance }: Props) => {
+  const [probes, setProbes] = useState<Probe[]>();
+  const [probesById, setProbesById] = useState<ProbeById>({});
+  const [selectedProbes, setSelectedProbes] = useState<Probe[]>([]);
+  const [probesToRemove, setProbesToRemove] = useState<number[]>([]);
+  const checks = selectedChecks();
+
+  const styles = useStyles(style);
+
+  const submitProbeUpdates = useCallback(async () => {
+    let newChecks: FilteredCheck[] = [];
+    // Short circuit remove action if all checks only have one probe
+    const checksHaveOneProbe = checks.every((check) => check.probes.length === 1);
+    if (action === 'remove' && checksHaveOneProbe) {
+      onDismiss();
+      onError('Operation canceled - all checks only have one probe');
+      return;
+    }
+    // Add or remove based on action
+    if (action === 'add') {
+      newChecks = checks.map((check) => {
+        // Using a Set to avoid duplicates
+        const newProbes = new Set([...check.probes, ...selectedProbes.map((p) => p.id!)]);
+        return { ...check, probes: [...newProbes] };
+      });
+    } else if (action === 'remove') {
+      // Filter out checks with only one probe
+      const checksWithMultipleProbes = checks.filter((check) => check.probes.length > 1);
+      newChecks = checksWithMultipleProbes.map((check) => {
+        const newProbes = check.probes.filter((p) => !probesToRemove.includes(p));
+        return { ...check, probes: newProbes };
+      });
+    }
+    try {
+      await instance.api?.bulkUpdateChecks(newChecks);
+      onDismiss();
+      clearSelections();
+      onSuccess();
+    } catch (error) {
+      onDismiss();
+      clearSelections();
+      onError(error.data.err);
+    }
+  }, [selectedProbes, checks, instance, action, probesToRemove, onDismiss, onError, onSuccess]);
+
+  const addOrRemoveProbe = useCallback(
+    (probe) => {
+      if (!selectedProbes.includes(probe)) {
+        setSelectedProbes((sp) => [...sp, probe]);
+      } else {
+        const newSelectedProbes = selectedProbes.filter((p) => p !== probe);
+        setSelectedProbes(newSelectedProbes);
+      }
+    },
+    [selectedProbes]
+  );
+
+  const addOrRemoveCommonProbe = useCallback(
+    (probe) => {
+      if (!probesToRemove.includes(probe)) {
+        setProbesToRemove((ptr) => [...ptr, probe]);
+      } else {
+        const newProbesToRemove = probesToRemove.filter((p) => p !== probe);
+        setProbesToRemove(newProbesToRemove);
+      }
+    },
+    [probesToRemove]
+  );
+
+  const clearSelections = () => {
+    setSelectedProbes([]);
+    setProbesToRemove([]);
+  };
+
+  const getProbes = useCallback(async () => {
+    const p = await instance.api!.listProbes();
```

Comment:
Is api for sure guaranteed to exist? I know the types are kind of wacky for that context